classes/timeslot_manager.php

```diff
@@ -447,6 +579,245 @@ public static function timeslot_html(int $observationid, int $slotid) {
             "start_time_formatted" => userdate($slotdata->start_time)
         ];
 
-        return $OUTPUT->render_from_template('mod_observation/confirm_message', $data);
+        return $OUTPUT->render_from_template($template, $data);
+    }
+
+    /**
+     * Creates a database record for a notification to send at a later date.
+     * @param int $observationid ID of the observation instance
+     * @param int $slotid ID of the timeslot to create a notification for
+     * @param int $userid ID of the user to create the notification for
+     * @param object $data Data containing details about the notification
+     */
+    public static function create_notification(int $observationid, int $slotid, int $userid, $data) {
+        global $DB;
+
+        $interval = $data->interval_amount;
+        $multiplier = $data->interval_multiplier;
+
+        if (!is_int($interval) || $interval < 1) {
+            throw new \coding_exception("Interval amount must be an integer that is greater than or equal to 1");
+        }
+
+        if (!is_int($multiplier) || $multiplier < 1) {
+            throw new \coding_exception("Multiplier amount must be an integer that is greater than or equal to 1");
+        }
+
+        // Ensure user hasn't created too many notifications.
+        $currentnotifications = self::get_users_notifications($observationid, $userid);
+
+        if (count($currentnotifications) >= self::MAX_NOTIFICATIONS) {
+            throw new \moodle_exception(get_string('toomanynotifications', 'observation'));
+        }
+
+        $DB->insert_record('observation_notifications', [
+            'timeslot_id' => $slotid,
+            'time_before' => $interval * $multiplier
+        ]);
+    }
+
+    /**
+     * Obtains the notifications for a user in a given observation activity
+     * @param int $observationid ID of the observation instance
+     * @param int $userid ID of the user to get notifications for
+     * @return array
+     */
+    public static function get_users_notifications(int $observationid, int $userid) {
+        global $DB;
+
+        $sql = '
+            SELECT obn.id as notification_id, time_before, ot.observee_id as userid
+              FROM {observation_notifications} obn
+         LEFT JOIN {observation_timeslots} ot
+                ON obn.timeslot_id = ot.id
+             WHERE ot.obs_id = :obsid
+               AND ot.observee_id = :userid
+        ';
```

Comment:
I would put the quotes on the same line as the query text, makes it a little easier to read imo